### PR TITLE
refactor reindex_available_works_job & reindex_model_jobs to accept ids for works to index

### PR DESCRIPTION
**Pull request by mankind**
We need to be able to pass the cnames of tenants to mint doi to each requeue of the job class. 
Secondly, **Tom W** wants us to accept a few ids to mint doi for to test it is working on demo sites, that way we don't have to loop through all the works in a work type  just to test that doi minting can be triggered by the Sidekiq job unless we are running it against production.

In the test calling this:

    expect { described_class.perform_now(work.class.to_s, account.cname, limit: 1, options: options) }.to change { work.doi_status_when_public }.to eq "findable"

 always  returns 
 
    expect X to have changed to eq "findable", but did not change 
 
Hence you will see that I used:

     described_class.perform_now(work.class.to_s, account.cname, limit: 1, options: options)
     work.reload
     expect(work.doi_status_when_public).to eq "findable"

### CODE REVIEW

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 1]
```diff
@@ -1,43 +1,38 @@
...
   class ReindexModelJob < ApplicationJob
-    queue_as :reindex
-
     rescue_from Hyrax::DOI::DataCiteClient::Error, Ldp::Gone, Ldp::HttpError, RSolr::Error::Http, RSolr::Error::ConnectionRefused do |exception|
       Rails.logger.debug exception.inspect
       retry_job(wait: 5.minutes)
     end
 
-    def perform(klass, cname, limit: 25, page: 1, cname_doi_mint: [])
+    def perform(klass, cname, limit: 25, page: 1, options: {})
```
Comment (thenapking) [inline app/jobs/hyku_addons/reindex_model_job.rb:16]: There are no tests updated for this class.

File: app/jobs/hyku_addons/reindex_available_works_job.rb [hunk 2]
```diff
@@ -21,7 +29,8 @@ def reindex(cname)
           next unless klass.count.positive?
 
           Rails.logger.debug "==== Queue ReindexModelJob for #{klass} ==="
-          HykuAddons::ReindexModelJob.perform_later(model_class, cname)
+          HykuAddons::ReindexModelJob.perform_later(model_class, cname, options: { cname_doi_mint: @cname_doi_mint }) if @cname_doi_mint.include?(cname)
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_available_works_job.rb:32]: I know that Hyrax/Hyku do this a lot, add extra conditions at the end of lines for `if/else`, but it woudl be clearer if we instead laid out a concise condition: 

```
options = @cname_doi_mint.include?(cname) ? { options: { cname_doi_mint: @cname_doi_mint } } : {}

HykuAddons::ReindexModelJob.perform_later(model_class, cname, options) 
```

File: app/jobs/hyku_addons/reindex_available_works_job.rb [hunk 3]
```diff
@@ -1,8 +1,16 @@
...
+# call with
+# HykuAddons::ReindexAvailableWorksJob.perform_now(['repo.hyku.docker'], cname_doi_mint: ["repo.hyku.docker"] )
+#
+# HykuAddons::ReindexAvailableWorksJob.perform_now(['repo.hyku.docker'])
+
 module HykuAddons
   class ReindexAvailableWorksJob < ApplicationJob
     # cnames is an array
-    def perform(cnames)
+    def perform(cnames, cname_doi_mint: [])
```
[the remark this answers is not in the document]
Comment (mankind) [reply app/jobs/hyku_addons/reindex_available_works_job.rb:11]: Because like time we spoke last time, the test was testing perform and I will refactor to test the parameters passed in but even it that it does not give me any confidence that testing that the right parameters were passed means the job will function as expected.
This is why i have not updated the test.

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 4]
```diff
@@ -1,43 +1,38 @@
...
-        new_page_count = page.to_i + 1
-
-        # Re-enqueue
-        ReindexModelJob.perform_later(klass, cname, page: new_page_count)
-      end
-
-      Rails.logger.debug "=== Completed reindex of #{klass} in #{cname} ==="
+      offset = (page - 1) * @limit
+      work_class = klass.constantize
+      fetch_work_using_ids(work_class) if @use_work_ids.present?
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:29]: Can we use a more immediately apparent condition?

```
if @use_work_ids.present?
    fetch_work_using_ids(work_class) 
else
    reindex_and_mint_work(work_class, offset) 
end
```

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 5]
```diff
@@ -51,9 +46,37 @@ def reindex_works(works)
...
+        Rails.logger.debug "=== Starting to reindex #{@klass} in #{@cname} ==="
+
+        # with calling to_a it always returns true, even when no records found
+        reindex_works(works)
+
+        new_page_count = @page.to_i + 1
+
+        # Re-enqueue
+        ReindexModelJob.perform_later(@klass, @cname, @limit, page: new_page_count, cname_doi_mint: @cname_doi_mint)
+        sleep(0.5)
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:78]: I'm not sure we should be adding sleep's to the jobs, this seems like a smell to me.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 1 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb:16]: So what is your suggestion @thenapking

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 6]
```diff
@@ -51,9 +46,37 @@ def reindex_works(works)
...
+
+        Rails.logger.debug "=== Starting to reindex #{@klass} in #{@cname} ==="
+
+        # with calling to_a it always returns true, even when no records found
+        reindex_works(works)
+
+        new_page_count = @page.to_i + 1
+
+        # Re-enqueue
+        ReindexModelJob.perform_later(@klass, @cname, @limit, page: new_page_count, cname_doi_mint: @cname_doi_mint)
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:77]: You dont need to add  new variable for the updated page count, the code is more readable without it: 

```ReindexModelJob.perform_later(@klass, @cname, @limit, page: @page.to_i + 1, cname_doi_mint: @cname_doi_mint)```

**Review by prdanelli — CHANGES_REQUESTED** [review]

File: app/jobs/hyku_addons/reindex_available_works_job.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (mankind) [reply app/jobs/hyku_addons/reindex_available_works_job.rb:32]: What do you think about just: 

                HykuAddons::ReindexModelJob.perform_later(model_class, cname, options: { cname_doi_mint: @cname_doi_mint })
                
This is because @cname_doi_mint always returns a fault of empty array

[the remark this answers is not in the document]
Comment (prdanelli) [reply app/jobs/hyku_addons/reindex_available_works_job.rb:32]: I said something very similar @thenapking. but as the `options` hash is set to a blank hash by default, if could be a ternary. 

```options = @cname_doi_mint.include?(cname) ? { options: { cname_doi_mint: @cname_doi_mint } } : {}

HykuAddons::ReindexModelJob.perform_later(model_class, cname, options) ```

File: app/jobs/hyku_addons/reindex_available_works_job.rb [hunk 7]
```diff
@@ -1,8 +1,16 @@
...
+      @cname_doi_mint = cname_doi_mint
+
       cnames.each do |name|
```
[the remark this answers is not in the document]
Comment (mankind) [reply app/jobs/hyku_addons/reindex_available_works_job.rb:14]: Ok , noted will make the change, thanks

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 5 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb:78]: I am happy to remove but what makes it a smell

Comment (thenapking) [reply app/jobs/hyku_addons/reindex_model_job.rb:78]: Especially not after asking a job to `perform_later`.

Comment (prdanelli) [reply app/jobs/hyku_addons/reindex_model_job.rb:78]: Just adding in sleep within a background job. A job should ideally just perform a single task, and if there are more tasks, each should be delegated. So having a sleep here, would imply that the job is doing a lot of work that needs to be broken up to prevent the site's performance from being hindered.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 6 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb:77]: I can make that change though i don't find your change easier to process if glancing quickly through the code compared to the current code.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 4 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb:29]: I thought you loved saving lines, was done with you in mind.

I will change it though/

File: app/jobs/hyku_addons/reindex_available_works_job.rb (hunk 2 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_available_works_job.rb:32]: I think this will do:
               HykuAddons::ReindexModelJob.perform_later(model_class, cname, options: { cname_doi_mint: @cname_doi_mint })

@cname_doi_mint is always an empty array, so I am not sure we need the check here:
    options = @cname_doi_mint.include?(cname) ? { options: { cname_doi_mint: @cname_doi_mint } } : {}

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 4 above)
Comment (prdanelli) [reply app/jobs/hyku_addons/reindex_model_job.rb:29]: hahaha. I do like saving lines, but clarity comes first  :D

I think its just this pattern that Hyrax uses, with these kind of end of line conditions for if/else... IMO they are not easier to read.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 1 above)
Comment (thenapking) [reply app/jobs/hyku_addons/reindex_model_job.rb:16]: This class calls save on some works, presumably in order to trigger reindexing them.  We could at the very least check that the correct works have been updated.  We can also check that the work's attribute `doi_status_when_public` gets set to `findable` if the work has a doi. 
As there are now two options that can be passed in to this job, I would expect there to be at least two test scenarios covering the use of each of the options.

File: app/jobs/hyku_addons/reindex_available_works_job.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (thenapking) [reply app/jobs/hyku_addons/reindex_available_works_job.rb:11]: Insofar as all this job does is spawn other jobs, the only thing I would expect to see in an updated test is that the spawned jobs have the new parameter passed through.  That's not the case with `ReindexModelJob`.

File: app/jobs/hyku_addons/reindex_available_works_job.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (thenapking) [reply app/jobs/hyku_addons/reindex_available_works_job.rb]: Yep, this is clearer to read as well I think.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 5 above)
Comment (thenapking) [reply app/jobs/hyku_addons/reindex_model_job.rb]: I agree.  You enqueue a job to be processed later, then wait half a second, then finish writing to the log.  What /does/ happen in that half a second, and why is it so important to sleep before writing your final comment to the log? There should be no reason to wait to write to the log, and it indicates a fragility or failure elsewhere.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 6 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb]: I am making the change but @prdanelli going with your desk analogy, a method with 2 to 3 local variables is like a desk with 2 or 3 books on it being called cluttered. Sounds more like a preference of having nothing on the desk and not a case of cluttering.

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 8]
```diff
@@ -1,43 +1,55 @@
 # frozen_string_literal: true
+
+# call with
+#  use_work_ids an array is just for testing a few ids
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:4]: Could you tidy this comment up?

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 9]
```diff
@@ -1,43 +1,55 @@
...
-
-      # When the offset becomes too large, no records would be found
-      works =  klass.constantize.where("title_tesim:*").limit(limit).offset(offset)
-
-      # with calling to_a it always returns true, even when no records found
-      if works.to_a.any?
-        reindex_works(works)
+      offset = (page - 1) * @limit
+      work_class = klass.constantize
+      fetch_work_using_ids(work_class) if @use_work_ids.present?
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:29]: I realise Rubocop gave you a warning about method length, but it could be resolved by removing the `offset` variable: 
```
work_class = klass.constantize
if @use_work_ids.present?
      fetch_work_using_ids(work_class) 
else
     # The offset variable is documented by the method signature
      reindex_and_mint_work(work_class, (page - 1) * @limit) 
end
```

It's better to add  `#rubocop:disable Metrics/MethodLength` and have a normal condition than two end of line conditions for `if` and `unless`.

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 10]
```diff
@@ -1,43 +1,55 @@
...
+      reindex_and_mint_work(work_class, offset) unless @use_work_ids.present?
+    end
 
-        new_page_count = page.to_i + 1
+    private
 
-        # Re-enqueue
-        ReindexModelJob.perform_later(klass, cname, page: new_page_count)
+      def mint_doi(work)
+        return if work.doi.present?
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:36]: Could you add a break return under your guard statements? It will make the code easier to read.

**Review by prdanelli — COMMENTED** [review]
I left a couple of comments. I think if we can tidy them up just a little more then its perfect.

File: spec/jobs/reindex_model_job_spec.rb [hunk 11]
```diff
@@ -14,11 +29,10 @@
...
   end
 
-  it "#perform_later can enqueue model for reindex later" do
-    expect { described_class.perform_later(work.class.to_s, account.cname) }.to have_enqueued_job(described_class).exactly(:once)
-  end
-
-  it "#perform_now will enqueue model for reindex" do
-    expect { described_class.perform_now(work.class.to_s, account.cname) }.to have_enqueued_job(described_class).exactly(:once)
+  it "sets doi_status_when_public to findable" do
+    expect(work.doi).to be_empty
```
Comment (thenapking) [inline spec/jobs/reindex_model_job_spec.rb:33]: We don't need to test this, it is defined as empty on line 5.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 10 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb:36]: ok i will add.

File: spec/jobs/reindex_model_job_spec.rb [hunk 12]
```diff
@@ -14,11 +29,10 @@
...
+    described_class.perform_now(work.class.to_s, account.cname, limit: 1, options: options)
+    work.reload
+    expect(work.doi_status_when_public).to eq "findable"
```
Comment (thenapking) [inline spec/jobs/reindex_model_job_spec.rb:36]: You can combine this into one line, as it is standard in job tests to have to reload the object changed by the job.
````
expect(work.reload.doi_status_when_public).to eq "findable"
````

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 9 above)
Comment (mankind) [reply app/jobs/hyku_addons/reindex_model_job.rb:29]: @prdanelli one of the point of the Rubocops is to help to provide pointers to things to refactor. So i don't really think we should always defer to adding **#rubocop:disable Metrics/MethodLength** rather than actually refactor the code. Otherwise what is point of the rubocop.

I am not convinced of the benefits of this change vs what is there?

File: spec/jobs/reindex_model_job_spec.rb [hunk 13]
```diff
@@ -1,11 +1,26 @@
 # frozen_string_literal: true
 
 RSpec.describe HykuAddons::ReindexModelJob do
-  let(:account) { create(:account) }
-  let(:work) { create(:work) }
+  let(:account) { create(:account, cname: cname) }
+  let(:work) { create(:work, doi: []) }
+  let(:second_work) { create(:work, title: ["another work"], doi: []) }
```
Comment (thenapking) [inline spec/jobs/reindex_model_job_spec.rb:6]: This isn't referenced anywhere in the test so we don't need to create it.

File: spec/jobs/reindex_model_job_spec.rb [hunk 14]
```diff
@@ -1,11 +1,26 @@
...
+  let(:prefix) { "10.1234" }
+  let(:cname) { "123abc" }
+  let(:response_body) { File.read(HykuAddons::Engine.root.join("spec", "fixtures", "doi", "mint_doi_return_body.json")) }
+  let(:options) { { cname_doi_mint: [account.cname], use_work_ids: [work.id] } }
 
   before do
-    allow(work.class).to receive(:reindex_everything)
+    GenericWork.all.map(&:destroy)
```
Comment (thenapking) [inline spec/jobs/reindex_model_job_spec.rb:13]: Why do we have to destroy all GenericWorks? Why have any objects been created for this test that we don't want created?

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 10 above)
Comment (prdanelli) [reply app/jobs/hyku_addons/reindex_model_job.rb:36]: Thank you buddy.

File: spec/jobs/reindex_model_job_spec.rb (hunk 14 above)
Comment (prdanelli) [reply spec/jobs/reindex_model_job_spec.rb:13]: FYI, `GenericWork.destroy_all` would be a slightly nicer way of writing this.

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 1 above)
Comment (thenapking) [reply app/jobs/hyku_addons/reindex_model_job.rb:16]: Thanks for adding the two test classes! I think this vastly improves the change.

File: spec/jobs/reindex_model_job_spec.rb (hunk 14 above)
Comment (prdanelli) [reply spec/jobs/reindex_model_job_spec.rb:13]: I just checked to make sure ActiveFedora implemented it: 

```
2.7.4 :001 > GenericWork.methods.include?(:destroy_all)
 => true
```

Comment (mankind) [reply spec/jobs/reindex_model_job_spec.rb:13]: @thenapking that ought to be taken out, I was checking something locally and just forgot to take off.

File: spec/jobs/reindex_model_job_spec.rb (hunk 13 above)
Comment (mankind) [reply spec/jobs/reindex_model_job_spec.rb:6]: @thenapking if you mean **second_work** it  was used in the test with issues that i spoke about in the stand up  i removed, so it ought to be off and will be removed.

File: spec/jobs/reindex_model_job_spec.rb (hunk 11 above)
Comment (mankind) [reply spec/jobs/reindex_model_job_spec.rb:33]: @thenapking  despite I discovered sometimes the work in the test I removed that was using **second_work** still had doi set, so this was added just for me to be double sure the doi is empty or it fails.

Comment (mankind) [reply spec/jobs/reindex_model_job_spec.rb:33]: @thenapking  despite that, I discovered that sometimes for the work in the test I removed that was using **second_work** the doi was set, when ought not to be so this was added just for me to be double sure tat the point of triggering the job that he doi was indeed empty otherwise the test  should fail.

I am happy to remove though.

File: app/jobs/hyku_addons/reindex_model_job.rb [hunk 15]
```diff
@@ -1,43 +1,58 @@
...
-
     rescue_from Hyrax::DOI::DataCiteClient::Error, Ldp::Gone, Ldp::HttpError, RSolr::Error::Http, RSolr::Error::ConnectionRefused do |exception|
       Rails.logger.debug exception.inspect
       retry_job(wait: 5.minutes)
     end
 
-    def perform(klass, cname, limit: 25, page: 1, cname_doi_mint: [])
+    # rubocop:disable Metrics/MethodLength
+    def perform(klass, cname, limit: 25, page: 1, options: {})
+      options ||= options || { cname_doi_mint: [], use_work_ids: [] }
```
Comment (prdanelli) [inline app/jobs/hyku_addons/reindex_model_job.rb:16]: This can just be `options ||= { cname_doi_mint: [], use_work_ids: [] }`

File: app/jobs/hyku_addons/reindex_model_job.rb (hunk 10 above)
Comment (thenapking) [reply app/jobs/hyku_addons/reindex_model_job.rb:38]: If this is something you feel strongly about we can add EmptyLineAfterGuardClause to rubocop.yml and then that will apply this automatically.

**Review by prdanelli — APPROVED** [review]
